### PR TITLE
sql: support array casting

### PR DESCRIPTION
**Pull request by rkrishn7**
### Motivation

  * This PR adds a known-desirable feature.
     - Solves #18703

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/expr/src/scalar/func/impls/array.rs [hunk 1]
```diff
@@ -136,3 +137,67 @@ impl fmt::Display for CastArrayToString {
...
+        let dims = arr.dims().into_iter().collect::<Vec<ArrayDimension>>();
+
+        let casted_datums = arr
+            .elements()
+            .iter()
+            .map(|datum| self.cast_expr.eval(&[datum], temp_storage))
+            .collect::<Result<Vec<Datum<'a>>, EvalError>>()?;
+
+        Ok(temp_storage.make_datum(|packer| {
+            packer.push_array(&dims, casted_datums).unwrap();
```
Comment (jkosh44) [inline src/expr/src/scalar/func/impls/array.rs:174]: Can you switch this to an `expect`?

File: test/sqllogictest/arrays.slt [hunk 2]
```diff
@@ -852,3 +852,64 @@ SELECT ARRAY[ARRAY[customer.first_name], ARRAY[customer.zip], ARRAY[customer.id:
...
+
+query T
+SELECT b::text[] FROM array_t2;
+----
+{{t1,t2,t3}}
+
+query T
+SELECT b::text[] FROM array_t2;
+----
+{{t1,t2,t3}}
```
Comment (jkosh44) [inline test/sqllogictest/arrays.slt:887]: These look identical, unless I'm missing something.

File: test/sqllogictest/arrays.slt [hunk 3]
```diff
@@ -852,3 +852,64 @@ SELECT ARRAY[ARRAY[customer.first_name], ARRAY[customer.zip], ARRAY[customer.id:
 query error db error: ERROR: array_in not yet supported
 SELECT array_in('', 0, 0)
 ----
+
+# Array casting for column expressions
```
Comment (jkosh44) [inline test/sqllogictest/arrays.slt:856]: Can you add some tests of the following format:

- Us `pg_tyepof` to assert that the final type is correct.
- Error scenarios where casts fail.

**Review by jkosh44 — COMMENTED** [review]
Thanks, so much for the PR! I had a couple of comments.

File: test/sqllogictest/arrays.slt (hunk 2 above)
Comment (rkrishn7) [reply test/sqllogictest/arrays.slt]: 🤦🏾 Thanks for catching that!

File: test/sqllogictest/arrays.slt (hunk 3 above)
Comment (rkrishn7) [reply test/sqllogictest/arrays.slt:856]: Definitely, added.

**Review by jkosh44 — APPROVED** [review]
LGTM! Thanks again for the PR! You have one lint issue, but once that's resolved I'll merge.

**Pull request merged**